pyface/data_view/abstract_data_model.py

```diff
@@ -0,0 +1,272 @@
+    values provided by the data, but not with how the data is presented.
+
+    Row and column indices are represented by sequences (usually lists) of
+    integers, specifying the index at each level of the heirarchy.  The root
+    row and column are represented by empty lists.
+
+    Subclasses need to implement the ``get_column_count``,
+    ``can_have_children`` and ``get_row_count`` methods to return the number
+    of columns in a particular row, as well as the heirarchical structure of
+    the rows.  Appropriate observers should be set up on the underlaying data
+    so that the ``structure_changed`` event is fired when the values returned
+    by these methods would change.
+
+    Subclasses also have to implement the ``get_value`` and ``get_value_type``
+    methods.  These expect a row and column index, with root values treated
+    specially: the root row corresponds to the values which will be displayed
+    in the column headers of the view, and the root column corresponds to the
+    values which will be displayed in the row headers of the view.
+    The ``get_value`` returns an arbitrary Python object corresponding to the
+    cell being viewed, and the ``get_value_type`` should return an instance of
+    an ``AbstractValueType`` that adapts the raw value to the data channels
+    that the data view expects (eg. text, color, icons, editable value, etc.).
+    Implementations should ensure that the ``values_changed`` event fires
+    whenever the data, or the way the data is presented, is updated.
+
+
+    If the data is to be editable then the subclass should override the
+    ``set_value`` method.  It should attempt to change the underlying data as a
+    side-effect, and return True on success and False on failure (for example,
+    setting an invalid value).  If the underlying data structure cannot be
+    listened to internally (such as a numpy array or Pandas data frame), this
+    method should also fire the values changed event with appropriate values.
+    """
+
+    #: The index manager that helps convert toolkit indices to data view
+    #: indices.  This should be an IntIndexManager for non-hierarchical data
+    #: or a TupleIndexManager for hierarchical data.
+    index_manager = Instance(AbstractIndexManager)
+
+    #: Event fired when the structure of the data changes.
+    structure_changed = Event()
+
+    #: Event fired when value changes without changes to structure.  This
+    #: should be set to a 4-tuple of (start_row_index, start_column_index,
+    #: end_row_index, end_column_index) indicated the subset of data which
+    #: changed.  These end values are inclusive, unlike standard Python
+    #: slicing notation.
+    values_changed = Event()
+
+    # Data structure methods
+
+    @abstractmethod
+    def get_column_count(self, row):
+        """ How many columns in the row of the data view model.
+
+        The total number of columns in the table is given by the column
+        count of the root row.
```

Comment:
This is boiling down to a restriction coming from the toolkits.  If you ask a non-leaf row in Qt "how many columns do you have?" then Qt will only show that many columns for all the child rows, even if they say that they have more, when asked.

It may be better of thinking of this as answering the question "how many columns do you need for your children?"

Reply:
In that case, if I have a N-D array `data` of shape `(10, 9, 8, 7, 6, 5)`, I would be tempted to return `sum(data.shape[len(row):])` based on the reading of this docstring, but that would be wrong.

Reply:
"Total" may be the misleading word here, since it can mean both "sum" and "all".